app/javascripts/main/packageManager.js

```diff
@@ -1,255 +1,267 @@
-          if (!err) {
-            this.unnestPackageContents(paths.absolutePath, () => {
-              // Find out main file
-              fileUtils.readJSONFile(paths.absolutePath + "/package.json", (response, error) => {
-                var main;
-                if (response) {
-                  if (response.sn) { main = response.sn.main; }
-                  if (response.version) { component.content.package_info.version = response.version; }
-                }
-                if (!main) { main = "index.html"; }
-
-                component.content.local_url = "sn://" + paths.relativePath + "/" + main;
-                callback(component);
-
-                // Update mapping file
-                this.updateMappingFile(component.uuid, paths.relativePath);
-              });
-            });
+async function runTasks(webContents, syncTasks) {
+  const tries = 3;
+  while (syncTasks.length > 0) {
+    try {
+      // Try to execute the task with up to three tries.
+      for (let triesAmount = 1; triesAmount <= 3; triesAmount++) {
+        try {
+          const taskData = syncTasks[0];
+          if (taskData.length === 1 && syncTasks.length > 1) {
+            // This is a single task, AKA an installation or deletion.
+            const oppositeTask = findOppositeTask(
+              syncTasks.slice(1),
+              taskData[0]
+            );
+            if (oppositeTask) {
+              // Found an opposite task. Delete it from the queue and do nothing.
+              syncTasks.splice(syncTasks.indexOf(oppositeTask), 1);
+              break;
+            }
+          }
+          await syncComponents(webContents, taskData);
+          break; // Everything went well, leave the loop.
+        } catch (error) {
+          if (triesAmount < tries) {
+            continue;
           } else {
-            // Unzip error
-            console.log("Unzip error for", component.content.name);
-            callback(component, { tag: "error-unzipping" });
+            throw error;
           }
-        });
-      } else {
-        // Download error
-        callback(component, { tag: "error-downloading" });
+        }
       }
-    });
-  }
-
-  /*
-    Maintains a JSON file which maps component ids to their installation location. This allows us to uninstall components
-    when they are deleted and do not have any `content`. We only have their uuid to go by.
-   */
-  updateMappingFile(componentId, componentPath) {
-    fileUtils.readJSONFile(MappingFileLocation, (response, error) => {
-      if (!response) response = {};
-
-      var obj = response[componentId] || {};
-      obj.location = componentPath;
-      response[componentId] = obj;
-
-      fs.writeFile(MappingFileLocation, JSON.stringify(response, null, 2), 'utf8', (err) => {
-        if (err) console.log("Mapping file save error:", err);
-      });
-    });
+    } catch (error) {
+      // TODO(baptiste): do something
```

Comment:
Anything need to be done here?

Reply:
I forgot to put a console.error there. I think I'll have to tackle error managing later on because it's also a UI thing and I'm not familiar enough with the codebase to tackle that right now